### PR TITLE
Proper handling $crate and local_inner_macros

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR introduces `HygineFrames` to store the macro definition/call site hierarchy in hyginee and when resolving `local_inner_macros` and `$crate`, we use the token to look up the corresponding frame and return the correct value.

See also: https://rustc-dev-guide.rust-lang.org/macro-expansion.html#hygiene-and-hierarchies

fixe #6890 and  #6788

r? @jonas-schievink

### CODE REVIEW

File: crates/hir_expand/src/hygiene.rs [hunk 1]
```diff
@@ -2,30 +2,96 @@
...
+        }
+        None
+    }
+}
+
+#[derive(Clone, Debug, Copy)]
+struct HygieneFrameId(usize);
+
+#[derive(Clone, Debug, Default)]
+struct HygieneFrames(Vec<HygieneFrame>);
```
Comment (jonas-schievink) [inline crates/hir_expand/src/hygiene.rs:69]: This can use `Arena` instead (then `HygieneFrameId` would become `Idx<HygieneFrame>`)

**Review by jonas-schievink — COMMENTED** [review]
Great!

File: crates/hir_expand/src/hygiene.rs (hunk 1 above)
Comment (edwin0cheng) [reply crates/hir_expand/src/hygiene.rs]: Yeap

**Comment by lnicola** [comment]
This has substantially slowed down the analysis:

image

Confirmed locally:

```
Database loaded 517.72ms, 284mi
Crates in this dir: 36
Total modules found: 576
Total declarations: 11153
Total functions: 8715
Item Collection: 9.38s, 93528mi
Total expressions: 239953                                                                                                                          
Expressions of unknown type: 2908 (1%)
Expressions of partially unknown type: 2578 (1%)
Type mismatches: 1200
Inference: 27.92s, 252342mi
Total: 37.30s, 345870mi

Database loaded 521.35ms, 284mi
Crates in this dir: 36
Total modules found: 576
Total declarations: 11146
Total functions: 8710
Item Collection: 7.08s, 74434mi
Total expressions: 238134                                                                                                                          
Expressions of unknown type: 2981 (1%)
Expressions of partially unknown type: 2640 (1%)
Type mismatches: 1193
Inference: 12.29s, 99177mi
Total: 19.37s, 173611mi
```

**Comment by edwin0cheng** [comment]
Maybe we could store `HygineFrames` in salsa ?

**Comment by jonas-schievink** [comment]
Unfortunately `ExpansionInfo` contains AST nodes, so persisting that in the database doesn't work (and would be a bad idea, since they're pretty heavy).

I think we should consider reverting this until we find a solution.

**Comment by edwin0cheng** [comment]
The original rustc implementation used TLS to store the hygiene information and I did try it and improve speed a lot. But I don't feel it is the right solution for us. 

I opened  https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/Hygiene.20Frames.20Information for the discussion.

**Pull request merged**